---
title: '[Figure 6] Differential Expression between Integration(+) and Integration(-) clones'
output:
  html_document:
    df_print: paged
---

This code takes the output of tximport_setup.Rmd, submits that gene expression data to DESeq2, and visualizes the output.


# Set up workspace

To ensure that this code is usable across a variety of platforms, the user-provided variables (e.g. the date of the run, the path to data files) are stored in the .csv file "userVars", which should be in the directory containing the directory that holds the .Rmd scripts.

## Import the variables you need
```{r}
# import .csv file with user paths and info written out
userVars <- read.delim("../userVars.csv", stringsAsFactors=FALSE, sep=",")

# print(userVars) # uncomment to include output in notebook to document variables in .html notebook output for this run

# user-determined variables used for generating filenames
currDate <- userVars[userVars$variable=='currentDate', 'path']
projDir <- userVars[userVars$variable=='projectDirectory', 'path']
myAlpha <- userVars[userVars$variable=='myAlpha', 'path']

pathToSetupData <- userVars[userVars$variable=='date_mostRecent_setup', 'path']
pathTo16pGeneList <- userVars[userVars$variable=='pathTo16pGeneList', 'path']
pathToSFARIgene <- userVars[userVars$variable=='pathToSFARIgene', 'path']
pathToSexGenes <- userVars[userVars$variable=='pathToSexGenes', 'path']


```

## Import needed data
```{r}
load( paste0(projDir, 'output/figure6/tximport_setup/', pathToSetupData, "/RData/", pathToSetupData, "_tximport_setup_setupVars.RData") )

```

## Make folders for this
```{r}
# name for files etc
expName <- 'deseq'

# make folder to contain experiment output

## for output
pathToAllProjectDir <- paste0(projDir, 'output/figure6/', expName)
dir.create(pathToAllProjectDir)

## main folder
pathToProjectDir <- paste0(projDir, 'output/figure6/', expName,'/', currDate)
dir.create(pathToProjectDir)

## make folder to contain non-RData output
pathToProjectOutput <- paste0(pathToProjectDir,'/other')
dir.create(pathToProjectOutput)

## make folder for RData vars
pathToRData <- paste0(pathToProjectDir,'/RData')
dir.create(pathToRData)

```


## load needed packages
```{r}
# Import Kristin's useful tools
library(DESeqAid)
library(DESeq2)
```

# Explore data - underlying structure in data that need to be batch corrected?

## Pre-cleaning Data

```{r}
# direct written outputs to designated project output folder
setwd(pathToProjectOutput)

# pre-filtering - filtering out empty rows. not at all necessary, but makes later computation faster.
dds_kal <- dds_kal[ rowSums(counts(dds_kal)) > 1, ]

# relevel - this explicitly tells DESeq to consider samples labeled 'WT' the controls
dds_kal$Genotype <- relevel(dds_kal$Genotype, ref = 'WT')

# Make a data frame containing the raw Count data stored in the DESeqDataSet
rawCounts <- data.frame( counts(dds_kal) )

```


# Deploy DESeq

## for gene
```{r}
# pre-filtering - filtering out empty rows. not at all necessary, but makes later computation faster.
dds_kal <- dds_kal[ rowSums(counts(dds_kal)) > 1, ]

# relevel - this explicitly tells DESeq to consider samples labeled 'N' the controls
dds_kal$PCR_integration <- relevel(dds_kal$PCR_integration, ref = 'N')

# DESeq
dds_deseq <- DESeq(dds_kal)

# direct written outputs to designated project output folder
setwd(pathToProjectOutput)

# perform DESeq analysis
deseqOutput <- deseq_contrasts(dds_deseq, myAlpha, c('PCR_integration', 'Y', 'N'), 'INTvCLEAN_gene', currDate, expName, geneList16p)

deseqOutput_df <- data.frame(deseqOutput)
deseqOutput_signif_df <- deseqOutput_df[which(deseqOutput_df$padj < as.numeric(myAlpha)),]
deseqOutput_signif_df <- deseqOutput_signif_df[order(deseqOutput_signif_df$padj),]
deseqOutput_signif_df_100 <- deseqOutput_signif_df[c(1:100),]

# only look at 16p genes
deseqOutput_16p<- subset(deseqOutput_df, row.names(deseqOutput_df) %in% geneList16p[,1] )

print('16p gene report:')
print(paste0('Number of 16p genes total: ', nrow(deseqOutput_16p)))
print(paste0('Number of DE 16p genes: ', nrow(deseqOutput_16p[deseqOutput_16p$padj < myAlpha,])))

# save data tables
setwd(pathToProjectOutput)
write.table(deseqOutput_signif_df , file = 'DEGenes.txt', quote = FALSE)
write.table(deseqOutput_signif_df_100, file = 'DEGenes_top100.txt', quote = FALSE)

setwd(pathToRData)
save( deseqOutput_signif_df_100, file = 'DEGenes_top100.RData')

# save DE gene names
setwd(pathToRData)
write.table(row.names(deseqOutput_signif_df) , file = 'namesOfDEGenes.txt', quote = FALSE, row.names=FALSE, col.names=FALSE)
namesOfDEGenes <- row.names(deseqOutput_df[which(deseqOutput_df$padj < myAlpha),])
save(namesOfDEGenes, file = 'namesOfDEGenes.RData')

```


## How much does that gene list overlap with putative POU5F1 targets?
https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-9-269
Used lists from 12864_2007_1462_MOESM17_ESM.xls 

Summary: 1-5 genes overlap between our DE list and those put forward in paper. Only a small fraction of genes. However, DE list might be too small due to inconsistency in amount of integration (or maybe would change if we used shp_transcript instead of PCR_integration)
```{r}
# import target list
oct4Targ <- read.delim("~/Dropbox/Projects/cnv16p/2017/resourcePaper/phenotype/integrationFig/kallisto/DESeq_20181009/20181012_oct4Targ_compare.csv")

oct4Targ_sharov <- deseqOutput_signif_df[ row.names(deseqOutput_signif_df) %in% oct4Targ[oct4Targ$Sharov == 1,'Gene_Upper'], ]
oct4Targ_matoba <- deseqOutput_signif_df[ row.names(deseqOutput_signif_df) %in% oct4Targ[oct4Targ$Matoba == 1,'Gene_Upper'], ]
oct4Targ_loh <- deseqOutput_signif_df[ row.names(deseqOutput_signif_df) %in% oct4Targ[oct4Targ$Loh == 1,'Gene_Upper'], ]


print('What Oct4 target genes appear in our DE gene list?')
print('Sharov N:')
print(nrow(oct4Targ_sharov))
print(paste0('Percent of list: ', (nrow(oct4Targ_sharov)/nrow(oct4Targ[oct4Targ$Sharov == 1,])) * 100 ))
print('Matoba N:')
print(nrow(oct4Targ_matoba))
print(paste0('Percent of list: ', (nrow(oct4Targ_matoba)/nrow(oct4Targ[oct4Targ$Matoba == 1,])) * 100 ))
print('Loh N:')
print(nrow(oct4Targ_loh))
print(paste0('Percent of list: ', (nrow(oct4Targ_loh)/nrow(oct4Targ[oct4Targ$Loh == 1,])) * 100 ))

# save results to .csv
setwd(pathToProjectOutput)
write.csv(oct4Targ_sharov[which(oct4Targ_sharov$padj < myAlpha),], file = 'Sharov_overlap.csv')
write.csv(oct4Targ_matoba[which(oct4Targ_matoba$padj < myAlpha),], file = 'Matoba_overlap.csv')
write.csv(oct4Targ_loh[which(oct4Targ_loh$padj < myAlpha),], file = 'Loh_overlap.csv')

```



# Visualize data

## Volcano plot of data
```{r}
# plot Volcano Plot
makeVolcanoPlot(deseqOutput_df, 
                geneList16p[,1], 
                '16p Region Gene', 
                paste0(currDate, ' ', expName), 
                'Log2(Shrunken Fold Change)', 
                '-Log10(Adjusted P-Value)',
                currDate,
                'volcanoPlot')

```



# 2019.06.04 // Quantified measure of extreme expression
```{r}

# size factor normalization
dds_kal <- estimateSizeFactors(dds_kal)
normalized_counts <- counts(dds_kal, normalized=TRUE)

# Integration severity across all DE genes genes

library(dplyr)

## prepare data frame of just DE genes
normDataInDf <- data.frame(normalized_counts)
just_de <- normDataInDf[row.names(normDataInDf) %in% row.names(deseqOutput_signif_df),] # MAKE SURE SIGNIF DE LIST HAS SOX13 LIKE IT SHOULD

probeMeans <- apply( just_de, 1, mean ) 
probeSd <- apply( just_de, 1, sd ) 

## calculate z-scores
zScores <- sweep(sweep(just_de, 1, probeMeans), 1, probeSd, FUN='/')

## which are down in Int+? (flip these signs)
downInInt <- row.names(deseqOutput_signif_df[which(deseqOutput_signif_df$log2FoldChange < 0),])
zScores[which(row.names(zScores) %in% downInInt),] <- zScores[which(row.names(zScores) %in% downInInt),]*-1

## calculate expression extremity
sample_zScore_sum <- data.frame(colSums(zScores))
sample_zScore_sum$integration_status <- 'clean'
sample_zScore_sum[which(row.names(sample_zScore_sum) %in% intClones),'integration_status'] <- 'integration_pos_by_pcr'
sample_zScore_sum$longID <- pull(metadata[match(row.names(sample_zScore_sum), metadata$DESeqAnalysisID), 'Line'])


```

## correlation of plas_* and (control?) with severity score
```{r}

## plasmid names
plasNames <- grep("plas",row.names(normDataInDf), value=TRUE)
namesToPlot <- c(plasNames, 'POU5F1', 'KIF22', 'PAX6', 'TP53', 'LRCH3', 'LIN28A', 'MYCL', 'KLF4', 'SOX2') # LRCH3 highly expressed and not DE
plasData <- t(normDataInDf[which(row.names(normDataInDf) %in% namesToPlot),])
plasData <- plasData[match(row.names(plasData), row.names(sample_zScore_sum)),]

## add in shp info for that
sample_zScore_sum_withExprs <- cbind(sample_zScore_sum, plasData)
sample_zScore_sum_withExprs$allPOU5F1 <- sample_zScore_sum_withExprs$POU5F1 + sample_zScore_sum_withExprs$plas_shp
sample_zScore_sum_withExprs$allSOX2 <- sample_zScore_sum_withExprs$SOX2 + sample_zScore_sum_withExprs$plas_hsk
sample_zScore_sum_withExprs$allKLF4 <- sample_zScore_sum_withExprs$KLF4 + sample_zScore_sum_withExprs$plas_hsk
sample_zScore_sum_withExprs$allMYCL <- sample_zScore_sum_withExprs$MYCL + sample_zScore_sum_withExprs$plas_hul
sample_zScore_sum_withExprs$allLIN28A <- sample_zScore_sum_withExprs$LIN28A + sample_zScore_sum_withExprs$plas_hul


## function useful for extracting correlation scors
corrFunc <- function(var1, var2, data) {
  results <- data.frame(var1 = c(), var2=c(), estimate=c(), p.value = c(), statistic = c(), method = c())
  for (m in c("pearson", "kendall")){
    print(paste0('Calculating correlation using method ', m))
    result = cor.test(data[,var1], data[,var2], method = m)
    newRow <- data.frame(var1, var2, result[c("estimate","p.value","statistic","method")], stringsAsFactors=FALSE)
    results <- rbind(results, newRow)
  }
  return(results)
  #return()
}


## loop through all genes
setwd('~/Desktop/20190814_corrPlots')
for (geneToCompare in names(sample_zScore_sum_withExprs)[4:length(sample_zScore_sum_withExprs)] ){
  
  # data we'll use
  scores <- sample_zScore_sum_withExprs[,1]
  exprs <- sample_zScore_sum_withExprs[geneToCompare][,1]
  
  # regression
  myLinReg <- lm(formula = exprs ~ scores) #myLm$coefficients[1], myLm$coefficients[2]
  
  # make plotData
  plotData <- data.frame(scores = scores, exprs = exprs)

  # ggplot scatterplot
  l_withError <- ggplot(plotData, aes(x=scores, y=exprs)) + 
    geom_point() +
    geom_smooth(method=lm) +
    theme_minimal() +
    ggtitle(paste0('Integration score vs. expression level for gene ', geneToCompare))
  
  ggsave(filename = paste0('',geneToCompare,'.pdf'),
         plot = l_withError,
         device = 'pdf',
         path = '~/Desktop/20190814_corrPlots',
         width = 20,
         height = 20,
         units = 'cm')
  
  l_noError <- ggplot(plotData, aes(x=scores, y=exprs)) + 
    geom_point() +
    geom_smooth(method=lm, se=FALSE) +
    theme_minimal() +
    ggtitle(paste0('Integration score vs. expression level for gene ', geneToCompare))
  
    ggsave(filename = paste0('',geneToCompare,'_noConfInt.pdf'),
         plot = l_withError,
         device = 'pdf',
         path = '~/Desktop/20190814_corrPlots',
         width = 20,
         height = 20,
         units = 'cm')

  print(l_withError)
  print(l_noError)
  
  print(paste0('### TESTING CORRELATION FOR ', geneToCompare, '###'))
  
  corrs <- corrFunc('scores', 'exprs', plotData)
  
  write.csv( corrs, file = paste0('correlations_scoreVgene_', geneToCompare, '.csv'), quote = FALSE )

  
  # plot
  # plot(scores,exprs,
  #      main = paste0('Z-score Severity Score vs. \nRlog Normed/Batch Corrected Expression of ', geneToCompare),
  #      xlab = 'Z-score based severity score',
  #      ylab = 'Expression')
  # abline(myLinReg$coefficients[1], myLinReg$coefficients[2])
  # 
  # print(paste0('### TESTING FOR ', geneToCompare, '###'))
  # write( cor.test(scores, exprs, method = c("pearson", "kendall", "spearman")) )

  
}


```


## do that agan but with 16p region genes
```{r}

## prepare data frame of just DE genes
just_de_16p <- assay(dds_kal_vst_bc)
just_de_16p <- just_de[row.names(just_de) %in% geneList16p[geneList16p[,1] %in% row.names(deseqOutput_signif_df),1],] # MAKE SURE SIGNIF DE LIST HAS SOX13 LIKE IT SHOULD

probeMeans_16p <- apply( just_de_16p, 1, mean ) 
probeSd_16p <- apply( just_de_16p, 1, sd ) 

## calculate z-scores
zScores_16p <- sweep(sweep(just_de_16p, 1, probeMeans), 1, probeSd, FUN='/')

## which are down in Int+? (flip these signs)
downInInt_16p <- row.names(deseqOutput_signif_df[which(deseqOutput_signif_df$log2FoldChange < 0),])
zScores_16p[which(row.names(zScores_16p) %in% downInInt_16p),] <- zScores_16p[which(row.names(zScores_16p) %in% downInInt_16p),]*-1

## calculate expression extremity
sample_zScore_sum_16p <- data.frame(colSums(zScores_16p))
sample_zScore_sum_16p$integration_status <- 'clean'
sample_zScore_sum_16p[which(row.names(sample_zScore_sum_16p) %in% intClones),'integration_status'] <- 'integration_pos_by_pcr'
sample_zScore_sum_16p$longID <- pull(metadata[ match(row.names(sample_zScore_sum_16p), metadata$DESeqAnalysisID) , 'Line'])

```


# HOUSEKEEPING
```{r}

setwd(pathToRData)
sessionInfo()
save.image(file=paste0(currDate, '_', expName, 'FinalAnalysis.RData'))

```
















