---
title: "[Figure 6] Prepare variables for analysis of transcriptional effects of integration status"
output: html_notebook
---

This code prepares variables for differential expression analysis (deseq.Rmd), and to explore the data via Principal Components Analysis clustering. Note that it takes in kallisto-aligned files only.

The sequences of the episomal expression vectors suspected of integration are available on the Addgene website:

https://www.addgene.org/27077/
https://www.addgene.org/27078/
https://www.addgene.org/27080/

# Set up Workspace

To ensure that this code is usable across a variety of platforms, the user-provided variables (e.g. the date of the run, the path to data files) are stored in the .csv file "userVars", which should be in the directory containing the directory that holds the .Rmd scripts.

## Import the variables you need
```{r}
# import .csv file with user paths and info written out
userVars <- read.delim("../userVars.csv", stringsAsFactors=FALSE, sep=",")

# print(userVars) # uncomment to include output in notebook to document variables in .html notebook output for this run

# user-determined variables used for generating filenames
currDate <- userVars[userVars$variable=='currentDate', 'path']
projDir <- userVars[userVars$variable=='projectDirectory', 'path']
pathToKallisto <- userVars[userVars$variable=='pathToKallisto', 'path']
pathToCleanClones <- userVars[userVars$variable=='pathToCleanClones', 'path']
pathToMetadata <- userVars[userVars$variable=='pathToMetadata', 'path']
pathToPCRIntegrationData<- userVars[userVars$variable=='pathToPCRIntegrationData', 'path']
pathTo16pGeneList <- userVars[userVars$variable=='pathTo16pGeneList', 'path']

```

## Make folders for this
```{r}
# name for files etc
expName <- 'tximport_setup'

# make folder to contain experiment output

## create folder for output
pathToAllProjectDir <- paste0(projDir, 'output/figure6/', expName)
dir.create(pathToAllProjectDir, recursive=TRUE)

## create folder for today's run
pathToProjectDir <- paste0(projDir, 'output/figure6/', expName,'/', currDate)
dir.create(pathToProjectDir)

## make folder to contain non-RData output
pathToProjectOutput <- paste0(pathToProjectDir,'/other')
dir.create(pathToProjectOutput)

## make folder for RData vars
pathToRData <- paste0(pathToProjectDir,'/RData')
dir.create(pathToRData)

```


# Import useful things

## Import 16p gene list
```{r}

# Import 16p Gene List
load(pathTo16pGeneList) 

```

## Import of metadata
```{r}
# Load metaData
library(readr)
metadata <- read_csv(pathToMetadata, na = "NA")

# Make sampleTable to instruct DESeqDataSetFromHTSeqCount how to import data
## sampleTable = rows are samples; 
## columns are (second) file names, (first) count file generated by htseq, (the rest) metadata
## put IDs in first two rows (since IDs are sample names *and* file names)
sampleTable_withUND <- data.frame( metadata[,c('DESeqAnalysisID', 'h5_FileName',
                                               colnames(metadata)[!colnames(metadata) 
                                                                  %in% c('ASD','Origin')] ) ] )

# trim out unwanted samples from sampleTable
sampleTable <- sampleTable_withUND[sampleTable_withUND$Genotype %in% c('WT', 'DEL', 'DUP'),] # remove anything with no file name in case of Excel spreadsheet error
sampleTable <- sampleTable[!(is.na(sampleTable$DESeqAnalysisID)),] # remove anything with no file name in case of Excel spreadsheet error

# add factor for high/low IQ
sampleTable$IQbin <- 'Unknown'
sampleTable[sampleTable$IQ > 75 & !is.na(sampleTable$IQ),'IQbin'] <- 'Above75'
sampleTable[sampleTable$IQ < 75 & !is.na(sampleTable$IQ),'IQbin'] <- 'Below75'

# declare all rows except for first two as factors
sampleTable[,-c(1,2)] <- lapply( sampleTable[,-c(1,2)], factor )
 # str(sampleTable) # uncomment to see that it worked

```

## Import Integration Status/MI data and create sampleTable_plus
```{r} 
# Import PCR Integration Data
IntData <- read_csv(pathToPCRIntegrationData)

# Combine IntData/DESeqIDs
comboIntDE <- merge(IntData, sampleTable[,c(names(sampleTable) %in% c('DESeqAnalysisID', 'Line'))], by='Line')

# Make sampleTable_plus
sampleTable_plus <- merge(sampleTable, comboIntDE, by='DESeqAnalysisID')

# declare all rows except for first two as factors
sampleTable_plus[,-c(1,2)] <- lapply( sampleTable_plus[,-c(1,2)], factor )

# clones that are Int+, Int-
intClones <- sampleTable_plus[sampleTable_plus$PCR_integration == 'Y','DESeqAnalysisID']
cleanClones <- sampleTable_plus[sampleTable_plus$PCR_integration == 'N','DESeqAnalysisID']

```



# Tximport to import data
```{r}
# import needed libraries
library(tximport)

# create file paths
files <- file.path(pathToKallisto, sampleTable_plus$tsv_FileName)
names(files) <- sampleTable_plus[ match(paste0(pathToKallisto,'/',sampleTable_plus$tsv_FileName), files), 'DESeqAnalysisID.1']

# create a transcript-to-gene-ID translator
library(EnsDb.Hsapiens.v86)
tx2gene <- transcripts(EnsDb.Hsapiens.v86, columns=c('tx_id', "symbol"), return.type="DataFrame")

# add plasmids to tx2gene
plasTx <- data.frame(tx_id = c('plas_hsk','plas_hul','plas_shp'),
                     symbol = c('plas_hsk','plas_hul','plas_shp'))

tx2gene <- rbind(tx2gene,plasTx)

# store kallisto data for use in DESeq
txi.kallisto <- tximport(files, type = "kallisto", txOut = FALSE, ignoreTxVersion = TRUE, tx2gene = tx2gene)

# store kallisto data as tpm matrix
txi.kallisto.tpm <- tximport(files, type = "kallisto", txOut = FALSE, ignoreTxVersion = TRUE, tx2gene = tx2gene, countsFromAbundance = 'scaledTPM')

```

# Make DESeqDataSet 

## (no normalization, no batch correction, gene-based) ("dds_kal")
```{r}
# import needed libraries
library(DESeq2)

# make data structure for raw data
dds_kal <- DESeqDataSetFromTximport(txi.kallisto, sampleTable_plus, ~PCR_integration)

# make data structure for tpm data
dds_kal_tpm <- DESeqDataSetFromTximport(txi.kallisto.tpm, sampleTable_plus, ~PCR_integration)

```

## (normalization, no batch correction, gene-based) ("dds_kal_vst") - for "before" PCA plots
```{r}
# Variance stabilizing transform
dds_kal_vst <- vst(dds_kal)

```


## Visualization of 16p11.2 abundance ("Christmas Tree Plot")
```{r}
# extract count data
vst_counts <- assay(dds_kal_vst)
vst_counts_16p <- vst_counts[which(row.names(vst_counts) %in% geneList16p[,1]), ]
metadata <- data.frame(metadata) # make sure metadata is a data frame

# Manipulate data table to add in factors needed for plotting
library(reshape2)
vst_counts_melt <- melt(vst_counts)
vst_counts_melt$Genotype <- metadata[match(vst_counts_melt$Var2, metadata$DESeqAnalysisID), 'Genotype']
colnames(vst_counts_melt) <- c('Gene', 'ID', 'value', 'Genotype')

library(dplyr)
vst_plot <- vst_counts_melt %>% group_by(Genotype,Gene) %>% summarize(myMean = mean(value)) 
vst_plot_2 <- dcast(vst_plot, Gene ~ Genotype)
vst_plot_2$gene16p <- ""
vst_plot_2$gene16p[which(vst_plot_2$Gene %in% geneList16p[,1])] <- 'TRUE'
vst_plot_2$gene16p <- factor(vst_plot_2$gene16p, levels = c("TRUE", ""))

# plot
library(ggplot2)
library(ggrepel)
ggplot(vst_plot_2, aes(x=WT, y=DEL, label = gene16p) ) + 
  geom_point(color = ifelse(!(vst_plot_2$gene16p == ""), "red", "grey50")) + theme_minimal() +
  geom_text_repel()

p <- ggplot() + 
  geom_point(data = vst_plot_2, aes(x=WT, y=DEL), color = "grey50") + 
  geom_point(data = vst_plot_2[which(vst_plot_2$Gene %in% geneList16p[,1]), ], aes(x=WT, y=DEL), color = "#BD4932") + 
  geom_abline(
      intercept=0,
      slope=1
    ) +
  theme_minimal() +
  geom_text_repel(data=vst_plot_2[which(vst_plot_2$Gene %in% geneList16p[,1]), ], aes(x=WT, y=DEL, label=Gene),
                    fontface="bold", color="#BD4932", force = 10)

```




## (normalization, batch correction, gene-based) ("dds_kal_vst_bc") - for WCGNA, "after" PCA plots
```{r}
# load needed libraries
library(limma)

# create new dds
dds_kal_vst_bc <- dds_kal_vst

## set up variables
limdesign <- model.matrix(~ PCR_integration + Sex + Genotype +  SeqBatch , data=colData(dds_kal))  # reminder: thing you care abt goes first
treatment.design <- limdesign[,1:2]
batch.design <- limdesign[,-(1:2)]

##'after' plots
assay(dds_kal_vst_bc) <- limma::removeBatchEffect(assay(dds_kal_vst_bc), design=treatment.design, covariates=batch.design) # remove batch

```


# How well are samples correlated?
```{r}
# install needed libraries
library('DESeqAid')

# plot sample correlation
sampSimilarityHeatmap(assay(dds_kal_vst), sampleTable_plus, c('Genotype', 'SubjectID', 'Sex', 'GrowBatch', 'Line.x', 'PCR_integration'), sampleTable_plus$DESeqAnalysisID)

```


# PCA plots of data

## Before batch correction

```{r, fig.width=12, fig.height=12}
# direct output to working directory
setwd(pathToProjectOutput)

# by genes
plotPCApcs_all <- plotPCA(dds_kal_vst, 'PCR_integration', returnData=TRUE)
plotPCApcs <- plotPCApcs_all[,c(1:2)]
colnames(plotPCApcs) <- c('PC1', 'PC2')

# plot PCs
distPCA(plotPCApcs, sampleTable_plus$PCR_integration, 'PCR Integration', 'PC1', 'PC2', paste0(currDate, ' ', expName ,'_names'))
distPCA(plotPCApcs, sampleTable_plus$transcripts_shp, 'shp_oct4Transcripts', 'PC1', 'PC2', paste0(currDate, ' ', expName,'_names'))
distPCA(plotPCApcs, sampleTable_plus$transcripts_hsk, 'hsk_klf4Transcripts', 'PC1', 'PC2', paste0(currDate, ' ', expName,'_names'))
distPCA(plotPCApcs, sampleTable_plus$Sex, 'Sex', 'PC1', 'PC2', paste0(currDate, ' ', expName,'_names'))
distPCA(plotPCApcs, sampleTable_plus$Genotype, 'Genotype', 'PC1', 'PC2', paste0(currDate, ' ', expName,'_names'))
distPCA(plotPCApcs, sampleTable_plus$SeqBatch, 'SeqBatch', 'PC1', 'PC2', paste0(currDate, ' ', expName,'_names'))
distPCA(plotPCApcs, sampleTable_plus$GrowBatch, 'GrowBatch', 'PC1', 'PC2', paste0(currDate, ' ', expName,'_names'))

```


## After batch correction 
```{r}
# direct output to working directory
setwd(pathToProjectOutput)

# by genes
plotPCApcs_all <- plotPCA(dds_kal_vst_bc, 'PCR_integration', returnData=TRUE)
plotPCApcs <- plotPCApcs_all[,c(1:2)]
colnames(plotPCApcs) <- c('PC1', 'PC2')

# plot PCAs
distPCA(plotPCApcs, sampleTable_plus$PCR_integration, 'PCR Integration', 'PC1', 'PC2', paste0(currDate, ' ', expName, 'afterBC')) #!! FIGURE
distPCA(plotPCApcs, sampleTable_plus$transcripts_shp, 'shp_oct4Transcripts', 'PC1', 'PC2', paste0(currDate, ' ', expName, 'afterBC'))
distPCA(plotPCApcs, sampleTable_plus$transcripts_hsk, 'hsk_klf4Transcripts', 'PC1', 'PC2', paste0(currDate, ' ', expName, 'afterBC'))
distPCA(plotPCApcs, sampleTable_plus$Sex, 'Sex', 'PC1', 'PC2', paste0(currDate, ' ', expName, 'afterBC'))
distPCA(plotPCApcs, sampleTable_plus$Genotype, 'Genotype', 'PC1', 'PC2', paste0(currDate, ' ', expName, 'afterBC'))
distPCA(plotPCApcs, sampleTable_plus$SeqBatch, 'SeqBatch', 'PC1', 'PC2', paste0(currDate, ' ', expName, 'afterBC'))
distPCA(plotPCApcs, sampleTable_plus$GrowBatch, 'GrowBatch', 'PC1', 'PC2', paste0(currDate, ' ', expName, 'afterBC'))

```



## After Batch Correction - plot more than first two principal components
```{r}
# direct output to working directory
setwd(pathToProjectOutput)

# calculate PCs
plotPCApcs_all <- plotPCA_pickPCs(dds_kal_vst_bc, 'PCR_integration', 'PC2', 'PC3', 2, 3, returnData=TRUE)
plotPCApcs <- plotPCApcs_all[,c(1:2)]
colnames(plotPCApcs) <- c('PC2', 'PC3')

# plot them
distPCA(plotPCApcs, sampleTable_plus$PCR_integration, 'PCR_Integration', 'PC2', 'PC3', paste0(currDate, ' ', expName, 'afterBC'))
distPCA(plotPCApcs, sampleTable_plus$Sex, 'Sex', 'PC2', 'PC3', paste0(currDate, ' ', expName, 'afterBC'))

```

# Update design for DESeq
```{r}
# update design
design(dds_kal) <- ~ Sex + Genotype +  SeqBatch + PCR_integration
```




# HOUSEKEEPING
```{r}
sessionInfo()
setwd(pathToRData)
save.image(file = paste0(currDate,'_',expName,'_setupVars.RData'))

```













