---
title: "[Figure 5] Differential Expression Analysis on Integration(-) clones"
output:
  html_notebook: default
  html_document:
    df_print: paged
  pdf_document: default
---

This code takes the output of setup.Rmd, submits that gene expression data to DESeq2, and visualizes the output.

# Set up workspace

To ensure that this code is usable across a variety of platforms, the user-provided variables (e.g. the date of the run, the path to data files) are stored in the .csv file "userVars", which should be in the directory containing the directory that holds the .Rmd scripts.

## Import the variables you need
```{r}
# import .csv file with user paths and info written out
userVars <- read.delim("../userVars.csv", stringsAsFactors=FALSE, sep=",")

# print(userVars) # uncomment to include output in notebook to document variables in .html notebook output for this run

# user-determined variables used for generating filenames
currDate <- userVars[userVars$variable=='currentDate', 'path']
projDir <- userVars[userVars$variable=='projectDirectory', 'path']
myAlpha <- as.numeric(userVars[userVars$variable=='myAlpha', 'path'])

pathToSetupData <- userVars[userVars$variable=='date_mostRecent_setup', 'path']
pathTo16pGeneList <- userVars[userVars$variable=='pathTo16pGeneList', 'path']
pathToSFARIgene <- userVars[userVars$variable=='pathToSFARIgene', 'path']
pathToSexGenes <- userVars[userVars$variable=='pathToSexGenes', 'path']

```


# Load needed variables
```{r}
load( paste0(projDir, 'output/figure5/setup/', pathToSetupData, "/RData/", pathToSetupData, "_setup_DESeqWorkspace.RData") )

load(pathTo16pGeneList)

sexGenes <- read.delim(pathToSexGenes, sep=",")
```

## Make folders for this
```{r}
# name for files etc
expName <- 'deseq'

# make folder to contain experiment output

## for output
pathToAllProjectDir <- paste0(projDir, 'output/figure5/', expName)
dir.create(pathToAllProjectDir)

## main folder
pathToProjectDir <- paste0(projDir, 'output/figure5/', expName,'/', currDate)
dir.create(pathToProjectDir)

## make folder to contain non-RData output
pathToProjectOutput <- paste0(pathToProjectDir,'/other')
dir.create(pathToProjectOutput)

## make folder for RData vars
pathToRData <- paste0(pathToProjectDir,'/RData')
dir.create(pathToRData)

```


# Differential Expression Analysis
```{r}
# Load needed variables
library(DESeqAid)

# direct written outputs to designated project output folder
setwd(pathToProjectOutput)

# DESeq
dds <- DESeq(myData)

# perform DESeq analysis
deseqOutput <- deseq_contrasts(dds, myAlpha, c('Genotype', 'DEL', 'WT'), 'WTvDEL_collapsed', currDate, expName, geneList16p)

deseqOutput_df <- data.frame(deseqOutput)
deseqOutput_signif_df <- deseqOutput_df[which(deseqOutput_df$padj < myAlpha),]

# save RData files with the above
setwd(pathToRData)
save(deseqOutput_df, file = 'deseqOutput_df_all.RData')
save(deseqOutput_signif_df, file = 'deseqOutput_df_signif_all.RData')

# only look at 16p genes
deseqOutput_16p<- subset(deseqOutput_df, row.names(deseqOutput_df) %in% geneList16p[,1] )

# save names of DE genes
setwd(pathToProjectOutput)
write.table(row.names(deseqOutput_signif_df) , file = 'namesOfDEGenes.txt', quote = FALSE, row.names=FALSE, col.names=FALSE)
namesOfDEGenes <- row.names(deseqOutput_df[which(deseqOutput_df$padj < myAlpha),])

setwd(pathToRData)
save(namesOfDEGenes, file = 'namesOfDEGenes.RData')

# print some reports
print(paste0('Number of genes DE with padj < ', myAlpha, ': ', nrow(deseqOutput_signif_df)) )

print('16p gene report:')
print(paste0('Number of 16p genes, total, in input list: ', nrow(deseqOutput_16p)))
print(paste0('Number of DE 16p genes: ', nrow(deseqOutput_16p[deseqOutput_16p$padj < myAlpha,])))


# Sex gene report
deseqOutput_sex<- subset(deseqOutput_df, row.names(deseqOutput_df) %in% sexGenes[,1] )
print('Sex gene report:')
print(paste0('Number of Sex genes, total, in input list: ', nrow(deseqOutput_sex)))
print(paste0('Number of DE Sex genes: ', nrow(deseqOutput_sex[deseqOutput_sex$padj < myAlpha,])))

# save expression data if you need it
myData_vst_bc_exprs <- assay(myData_vst_bc)

```


# Create files that Theo wanted (10/16/2019)
```{r}

# Combine DESeq output and expression data
myData_vst_bc_exprs

exprsAndDEOutput <- merge(myData_vst_bc_exprs, deseqOutput_16p[, c('log2FoldChange', 'padj')], by='row.names')

# Are those genes in our gene list?

## Import Theo's list of genes
theo16pList <- read.table("~/Desktop/20191016_newTheo16pList.csv", quote="\"", comment.char="")

## In DE list?
exprsAndDEOutput$theoList <- FALSE
exprsAndDEOutput[which(exprsAndDEOutput$Row.names %in% theo16pList[,1]), 'theoList'] <- TRUE

## What about raw data?
rawData <- data.frame(counts(myData))
rawData$theoList <- FALSE
rawData[which(row.names(rawData) %in% theo16pList[,1]), 'theoList'] <- TRUE

# Save
setwd('~/Desktop')
write.csv(exprsAndDEOutput, file='20191016_exprsAndDEOutput.csv', quote=FALSE)

```


# Visualize data

## Volcano plot of data
```{r}
setwd(pathToProjectOutput)

makeVolcanoPlot(deseqOutput_df, 
                geneList16p[,1], 
                '16p Region Gene', 
                paste0(currDate, ' ', expName), 
                'Log2(Shrunken Fold Change)', 
                '-Log10(Adjusted P-Value)',
                currDate,
                'volcanoPlot')

```

## Bar Plot
```{r}
setwd(pathToProjectOutput)

fcSortBarPlot_lfcShrink(deseqOutput, geneList16p, '16p Genes', 0.58, 'DE Genes using design = ~Sex + GrowBatch + Genotype', 30, 7)
fcSortBarPlot_lfcShrink(deseqOutput, geneList16p, '16p Genes', 0, 'DE Genes using design = ~Sex + GrowBatch + Genotype', 12, 5)


```


## Make heatmap
### Prep plotting data
```{r}

library(RColorBrewer)

# reorganize sampleTable_collapsed so F at the top of each category - order by two cols, Genotype and Sex
sampleTable_collapsed <- sampleTable_collapsed[with(sampleTable_collapsed, order(Genotype, Sex)), ]

# create plotting data - just DE genes
plotData <- assay(myData_vst_bc)
plot_de <- plotData[which(rownames(plotData) %in% namesOfDEGenes),]
plot_de <- plot_de[,match(sampleTable_collapsed$Line, colnames(plot_de))]

# create annotations
my_pheno_col <- data.frame(sampleTable_collapsed[,c('Genotype', 'Sex')])
  
# define annotation colors
annColors <- list(Genotype = c(DEL='#636363',WT='#bdbdbd'),
                  Sex = c(F = '#000000', M= '#bdbdbd') )

  
```


### Plot: Z-scores, VST-normalized, batch corrected expression
```{r}

# load needed libraries
library(pheatmap)

# Aug 13 2019: make special area
setwd(file.path(pathToProjectOutput, '/specialGraphs_08132019'))


# plot pheatmap

## Using colorbrewer
pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_BuPu_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="BuPu")))(100),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_YlOrRd_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="YlOrRd")))(100),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_YlOrBr_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="YlOrBr")))(100),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_Greens_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="Greens")))(100),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_Purples_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="Purples")))(100),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_PuRd_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="PuRd")))(100),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_PuBuGn_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="PuBuGn")))(100),
         scale='column',
         annotation_colors = annColors)


pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_PuBu_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="PuBu")))(100),
         scale='column',
         annotation_colors = annColors)


pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_RdPu_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="RdPu")))(100),
         scale='column',
         annotation_colors = annColors)


pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_BuGn_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="BuGn")))(100),
         scale='column',
         annotation_colors = annColors)


pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_cbrewer_GnBu_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = colorRampPalette(rev(brewer.pal(n = 7, name ="GnBu")))(100),
         scale='column',
         annotation_colors = annColors)


## Viridis
library(viridis)
pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_viridis_optionD_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = viridis(100, option = "D"),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_viridis_optionA_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = viridis(100, option = "A"),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_viridis_optionB_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = viridis(100, option = "B"),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_viridis_optionC_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = viridis(100, option = "C"),
         scale='column',
         annotation_colors = annColors)

## wes anderson palettes
install.packages("wesanderson")
library(wesanderson)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_wesanderson_zissou1_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = wes_palette("Zissou1", 100, type = "continuous"),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_wesanderson_rushmore1_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = wes_palette("Rushmore1", 100, type = "continuous"),
         scale='column',
         annotation_colors = annColors)


pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_wesanderson_royal1_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = wes_palette("Royal1", 100, type = "continuous"),
         scale='column',
         annotation_colors = annColors)


pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_wesanderson_royal2_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = wes_palette("Royal2", 100, type = "continuous"),
         scale='column',
         annotation_colors = annColors)


pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_wesanderson_darjeeling1_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = wes_palette("Darjeeling1", 100, type = "continuous"),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_wesanderson_darjeeling2_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = wes_palette("Darjeeling2", 100, type = "continuous"),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_wesanderson_fantasticfox1_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = wes_palette("FantasticFox1", 100, type = "continuous"),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_wesanderson_moonrise1_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = wes_palette("Moonrise1", 100, type = "continuous"),
         scale='column',
         annotation_colors = annColors)

pheatmap(t(plot_de), 
         annotation_row = my_pheno_col, 
         cluster_rows = FALSE, gaps_row = 13, 
         filename = 'degene_scaled_wesanderson_grandbudapest1_LineNames_4p8.pdf', 
         width = 30, 
         height = 4.8, 
         color = wes_palette("GrandBudapest1", 100, type = "continuous"),
         scale='column',
         annotation_colors = annColors)
```




# How much does this gene list intersect with SFARI gene list?

## Import SFARI gene list
```{r}
# import SFARI gene ASD risk gene list
sfariList <- read.csv(pathToSFARIgene, header=TRUE)
sfariList$gene.score <- as.factor(sfariList$gene.score)

```


## Merge our DE genes with SFARI list
```{r}

# merge the two
genes_annot_onlyRisk <- merge(deseqOutput_df, sfariList, by.x='row.names', by.y='gene.symbol')
genes_annot <- merge(deseqOutput_df, sfariList, by.x='row.names', by.y='gene.symbol', all=TRUE)

# order by padjusted
genes_annot_onlyRisk <- genes_annot_onlyRisk[order(genes_annot_onlyRisk$padj, decreasing=FALSE),]
genes_annot <- genes_annot[order(genes_annot$padj, decreasing=FALSE),]

# write to file
setwd(pathToProjectOutput)
write.table(genes_annot, paste0(currDate, '_deseqOutput_plusSFARI.csv'), quote=FALSE)
write.table(genes_annot_onlyRisk, paste0(currDate, '_deseqOutput_plusSFARI_sfariRiskOnly.csv'), quote=FALSE)
```


## Visualization of 16p11.2 abundance ("Christmas Tree Plot")
```{r}

# identify significantly DE DESeq Output Genes
deseqOutput_16p_sig <- deseqOutput_16p[which(deseqOutput_16p$padj < 0.05), ]

# Manipulate data table to add in factors needed for plotting
library(reshape2)
plotData_melt <- melt(plotData)
plotData_melt$Genotype <- sampleTable_collapsed[match(plotData_melt$Var2, sampleTable_collapsed$Line), 'Genotype']
colnames(plotData_melt) <- c('Gene', 'ID', 'value', 'Genotype')

library(dplyr)
plotData_plot <- plotData_melt %>% group_by(Genotype,Gene) %>% summarize(myMean = mean(value)) 
plotData_plot_2 <- dcast(plotData_plot, Gene ~ Genotype)
plotData_plot_2$gene16p <- ""
plotData_plot_2$gene16p[which(plotData_plot_2$Gene %in% geneList16p[,1])] <- "TRUE"
plotData_plot_2$gene16p <- factor(plotData_plot_2$gene16p, levels = c("TRUE", ""))

plotData_plot_2$gene16p <- plotData_plot_2$Gene
plotData_plot_2$gene16p <- factor(plotData_plot_2$gene16p, levels = c(geneList16p, ""))
plotData_plot_2$gene16p[which(!(plotData_plot_2$Gene %in% geneList16p[,1]))] <- ""

# plot

## make basic plot
library(ggplot2)
library(ggrepel)
ggplot(plotData_plot_2, aes(x=WT, y=DEL, label = gene16p) ) + 
  geom_point(color = ifelse(!(plotData_plot_2$gene16p == ""), "red", "grey50")) + theme_minimal() +
  geom_text_repel()

## add in layers representing
### Rd/Or color scheme: noDE = "#F28E2B", DE = "#E15759"
### Or/Sf: noDE = "#1b9e77", DE= "#d95f02" (colorbrewer approved)
### Pu/Gr: noDE = "#7570b3", DE= "#1b9e77" (colorbrewer approved)

genes16p_de_data <- plotData_plot_2[which(plotData_plot_2$Gene %in% row.names(deseqOutput_16p_sig)), ]
genes16p_notDE_data <- plotData_plot_2[which(plotData_plot_2$Gene %in% geneList16p[,1] & !(plotData_plot_2$Gene %in% row.names(deseqOutput_16p_sig)) ), ]

### both types of 16p genes
p <- ggplot() + 
  geom_point(data = plotData_plot_2, aes(x=WT, y=DEL), color = "#BAB0AC") + 
  geom_point(data = genes16p_notDE_data, aes(x=WT, y=DEL), color = "#7570b3", size=1) + 
  geom_point(data = genes16p_de_data, aes(x=WT, y=DEL), color = "#1b9e77", size=1) + 
  geom_abline(
      intercept=0,
      slope=1
    ) +
  theme_minimal() +
  geom_label_repel(data=genes16p_notDE_data, aes(x=WT, y=DEL, label=Gene), fontface="bold", color="#7570b3", force = 10, segment.size = 0.2, direction = "x", segment.color = "grey50", nudge_y = 16 - subset(genes16p_notDE_data, WT < 9)$WT) +
  geom_label_repel(data=genes16p_de_data, aes(x=WT, y=DEL, label=Gene), fontface="bold", color="#1b9e77", force = 20, segment.size = 0.2, direction = "y", segment.color = "grey50", nudge_x = 15 - subset(genes16p_de_data, WT < 10)$WT)

setwd('~/Desktop')
ggsave('xmasTreePlot_cleanClones_vst_bc_DE16pgenes_other16pgenes_PuGr_size1.pdf', p, device = 'pdf', height=20, width=25, unit = 'cm')

### just DE genes
p <- ggplot() + 
  geom_point(data = plotData_plot_2, aes(x=WT, y=DEL), color = "#BAB0AC") + 
  geom_point(data = genes16p_de_data, aes(x=WT, y=DEL), color = "#7570b3", size=1) + 
  geom_abline(
      intercept=0,
      slope=1
    ) +
  theme_minimal() +
  geom_label_repel(data=genes16p_de_data, aes(x=WT, y=DEL, label=Gene), fontface="bold", color="#7570b3", force = 20, segment.size = 0.2, direction = "y", segment.color = "grey50", nudge_x = 15 - subset(genes16p_de_data, WT < 10)$WT)

setwd('~/Desktop')
ggsave('xmasTreePlot_cleanClones_vst_bc_DE16pgenes_Rd_size1.pdf', p, device = 'pdf', height=20, width=25, unit = 'cm')

```



# 2019.06.05 // Quantified measure of extreme expression for Aditya
```{r}

# Integration severity across all DE genes genes

## prepare data frame of just DE genes
just_de <- assay(myData_vst_bc)
just_de <- just_de[row.names(just_de) %in% row.names(deseqOutput_signif_df),] # MAKE SURE SIGNIF DE LIST HAS SOX13 LIKE IT SHOULD

probeMeans <- apply( just_de, 1, mean ) 
probeSd <- apply( just_de, 1, sd ) 

## calculate z-scores
zScores <- sweep(sweep(just_de, 1, probeMeans), 1, probeSd, FUN='/')

## which are down in DEL? (flip these signs)
downInDEL <- row.names(deseqOutput_signif_df[which(deseqOutput_signif_df$log2FoldChange < 0),])
zScores[which(row.names(zScores) %in% downInDEL),] <- zScores[which(row.names(zScores) %in% downInDEL),]*-1

## calculate expression extremity
sample_zScore_sum <- data.frame(colSums(zScores))
sample_zScore_sum$Genotype
sample_zScore_sum$Genotype <- sampleTable_collapsed[match( row.names(sample_zScore_sum),
                                                         row.names(sampleTable_collapsed)),'Genotype']

setwd('~/Desktop')
write.csv(sample_zScore_sum, 'intNeg_WildTypeNessScore.csv')

```

## that but with just 16p genes
```{r}

# Integration severity across all DE genes genes

## prepare data frame of just DE genes
just_16p <- assay(myData_vst_bc)
just_16p <- just_16p[row.names(just_16p) %in% geneList16p[,1],] # MAKE SURE SIGNIF DE LIST HAS SOX13 LIKE IT SHOULD

probeMeans_16p <- apply( just_16p, 1, mean ) 
probeSd_16p <- apply( just_16p, 1, sd ) 

## calculate z-scores
zScores_16p <- sweep(sweep(just_16p, 1, probeMeans_16p), 1, probeSd_16p, FUN='/')

## calculate expression extremity
sample_zScore_sum_16p <- data.frame(colSums(zScores_16p))
sample_zScore_sum_16p$Genotype
sample_zScore_sum_16p$Genotype <- sampleTable_collapsed[match( row.names(sample_zScore_sum_16p),
                                                         row.names(sampleTable_collapsed)),'Genotype']

setwd('~/Desktop')
write.csv(sample_zScore_sum_16p, 'intNeg_WildTypeNessScore_16p.csv')

```

# HOUSEKEEPING
Run this block to save all your variables and store useful information about this run in the html file that accompanies this notebook.
```{r}
sessionInfo()

setwd(pathToRData)
save.image(file = paste0(currDate,'_',expName,'allVars.RData'))

```


